yarn-project/acir-simulator/src/avm/opcodes/external_calls.ts

```diff
@@ -0,0 +1,83 @@
+import { Fr } from '@aztec/foundation/fields';
+
+import { AvmContext } from '../avm_context.js';
+import { AvmMachineState } from '../avm_machine_state.js';
+import { AvmJournal } from '../journal/journal.js';
+import { Instruction } from './instruction.js';
+
+export class Call extends Instruction {
+  static type: string = 'CALL';
+  static numberOfOperands = 7;
+
+  constructor(
+    private /* Unused due to no formal gas implementation at this moment */ _gasOffset: number,
+    private addrOffset: number,
+    private argsOffset: number,
+    private argSize: number,
+    private retOffset: number,
+    private retSize: number,
+    private successOffset: number,
+  ) {
+    super();
+  }
+
+  // TODO(https://github.com/AztecProtocol/aztec-packages/issues/3992): there is no concept of remaining / available gas at this moment
+  async execute(machineState: AvmMachineState, journal: AvmJournal): Promise<void> {
+    const callAddress = machineState.readMemory(this.addrOffset);
+    const calldata = machineState.readMemoryChunk(this.argsOffset, this.argSize);
+
+    const avmContext = AvmContext.prepExternalCall(callAddress, machineState.executionEnvironment, journal);
+
+    const returnObject = await avmContext.call(calldata);
+    const success = !returnObject.reverted;
+
+    // We only take as much data as was specified in the return size -> TODO: should we be reverting here
+    const returnData = returnObject.output.slice(0, this.retSize);
+
+    // Write our return data into memory
+    machineState.writeMemory(this.successOffset, new Fr(success));
+    machineState.writeMemoryChunk(this.retOffset, returnData);
+
+    avmContext.mergeJournal();
+
+    this.incrementPc(machineState);
+  }
+}
+
+export class StaticCall extends Instruction {
+  static type: string = 'STATICCALL';
+  static numberOfOperands = 7;
+
+  constructor(
+    private /* Unused due to no formal gas implementation at this moment */ _gasOffset: number,
+    private addrOffset: number,
+    private argsOffset: number,
+    private argSize: number,
+    private retOffset: number,
+    private retSize: number,
+    private successOffset: number,
+  ) {
+    super();
+  }
+
+  async execute(machineState: AvmMachineState, journal: AvmJournal): Promise<void> {
+    const callAddress = machineState.readMemory(this.addrOffset);
+    const calldata = machineState.readMemoryChunk(this.argsOffset, this.argSize);
+
+    const avmContext = AvmContext.prepExternalStaticCall(callAddress, machineState.executionEnvironment, journal);
+
+    const returnObject = await avmContext.call(calldata);
```

Comment:
In principle it feels strange that an instruction ends up transitively using an Interpreter, and that an interpreter uses Instructions. I understand that conceptually you "switch context" and "start a new execution" but this kind of breaks the elegance of having instructions be basically just a lambda.

I'm a bit surprised that this works, but we could leave it like this until we find a better way (which likely involve a more complex design).

Reply:
I agree here, it feels a little recursive. But in my mind a nested instruction call is somewhat recursive, in this case the downstream call needs a completely fresh context to work within